include/picongpu/plugins/saxs/Saxs.hpp

```diff
@@ -0,0 +1,815 @@
+                        bufferSize_i,
+                        complex_X( 0.0 )
+                    );
+                    bufferSize = DataSpace< DIM1 >( bufferSize_i );
+                }
+                // Allocate amplitude buffer.
+                amplitude = pmacc::memory::makeUnique< ComplexBuffer >(
+                    bufferSize );
+                // Initialize, on device, its fields with zero.
+                amplitude->getDeviceBuffer( ).setValue( 0.0 );
+
+                // Set the q-space grid spacing.
+                q_step = ( q_max - q_min ) /
+                         precisionCast< float_X >( numVectors );
+
+                // Rank 0 creates the output directory.
+                pmacc::Filesystem< simDim > & fs =
+                    Environment< simDim >::get( ).Filesystem( );
+                if ( isMaster )
+                {
+                    fs.createDirectory( "saxsOutput" );
+                    fs.setDirectoryPermissions( "saxsOutput" );
+                }
+
+                // Chose the solver for populating a TmpField with the electron
+                // density (either the species density or the bound electron
+                // density).
+                using ElectronDensitySolver =
+                    typename traits::DetermineElectronDensitySolver<
+                    T_ParticlesType >::type;
+                // Output unit:
+                const float_64 amplitudeUnit = static_cast< float_64 >(
+                    FieldTmp::getUnit< ElectronDensitySolver >( )[ 0 ] ) *
+                    CELL_WIDTH_SI * CELL_HEIGHT_SI * CELL_DEPTH_SI
+                    * ELECTRON_RADIUS_SI;
+
+                // Initialize an object responsible for output writing.
+                dataWriter = pmacc::memory::makeUnique< SaxsWriter< float_X > >(
+                    pluginPrefix + "Output",
+                    fileExtension,
+                    "saxsOutput",
+                    outputLayout,
+                    compressionMethod,
+                    precisionCast< uint64_t >( numVectors ),
+                    q_step,
+                    amplitudeUnit
+                );
+                // Set the output period.
+                outputPeriod = pluginSystem::toTimeSlice( outputPeriod_s );
+            }
+        }
+
+
+        void pluginUnload( ) override
+        {
+            if ( !notifyPeriod.empty( ) )
+            {
+                CUDA_CHECK( cudaGetLastError( ) );
+            }
+        }
+
+
+        //! Collect amplitude data from each CPU on the master node.
+        void collectIntensityOnMaster( )
+        {
+            auto const totalNumVectors = numVectors.productOfComponents( );
+
+            amplitude->deviceToHost( );
+            __getTransactionEvent( ).waitForFinished( );
+
+            reduce(
+                nvidia::functors::Add( ),
+                amplitudeMaster.data( ),
+                amplitude->getHostBuffer( ).getBasePointer( ),
+                bufferSize_i,
```

Comment:
Here I think it is better to get the size from the buffer, since it's guaranteed to match the actual buffer, while a separate variable might become outdated.